crates/wit-parser/src/resolve.rs

```diff
@@ -293,32 +309,64 @@ impl Resolve {
+    pub fn push_dir(&mut self, path: &Path) -> Result<(PackageId, Vec<PathBuf>)> {
+        let top_pkg = UnresolvedPackageGroup::parse_dir(path)
             .with_context(|| format!("failed to parse package: {}", path.display()))?;
-        let (pkgs_ids, mut top_level_path_bufs) =
-            self.sort_unresolved_packages(vec![unresolved_top_level])?;
+        let deps = path.join("deps");
+        let mut deps = self
+            .parse_deps_dir(&deps)
+            .with_context(|| format!("failed to parse dependency directory: {}", deps.display()))?;
+        // Perform a simple topological sort which will bail out on cycles
+        // and otherwise determine the order that packages must be added to
+        // this `Resolve`.
+        let mut order = IndexSet::new();
+        let mut visiting = HashSet::new();
+        for pkg in &top_pkg.nested {
+            visit_group(&pkg, &deps, &mut order, &mut visiting, &pkg.source_map)?;
+        }
+        for pkg in deps.values().chain([&top_pkg]) {
+            for pkg in &pkg.nested {
+                visit_group(&pkg, &deps, &mut order, &mut visiting, &pkg.source_map)?;
+            }
+            visit_group(&pkg.main, &deps, &mut order, &mut visiting, &pkg.source_map)?;
+        }
+
+        // Using the topological ordering insert each package incrementally.
+        // Additionally note that the last item visited here is the root
+        // package, which is the one returned here.
+        let mut files = Vec::new();
+        let mut main = top_pkg;
+        for name in order {
+            if deps.contains_key(&name) {
+                let pkg = deps.remove(&name).unwrap_or_else(|| main.clone());
+                files.extend(pkg.source_map.source_files().map(|p| p.to_path_buf()));
+                for mut nested in pkg.nested {
+                    self.push(&mut nested)?;
+                }
+                self.push(&mut pkg.main.clone())?;
+            } else {
+                let cur_pkg = main.nested.iter_mut().find(|p| p.name == name);
+                if let Some(mut pkg) = cur_pkg {
+                    files.extend(pkg.source_files().map(|p| p.to_path_buf()));
+                    self.push(&mut pkg)?;
+                }
+            }
+        }
+
+        files.extend(main.source_map.source_files().map(|p| p.to_path_buf()));
+        main.main.source_map = main.source_map;
+        let pkgid = self.push(&mut main.main)?;
 
-        path_bufs.append(&mut top_level_path_bufs);
-        Ok((pkgs_ids, path_bufs))
+        return Ok((pkgid, files));
     }
 
-    fn parse_deps_dir(&mut self, path: &Path) -> Result<Vec<UnresolvedPackageGroup>> {
-        let mut unresolved_deps = Vec::new();
-        // If there's no `deps` dir, then there's no deps, so return the
-        // empty set.
+    fn parse_deps_dir(
+        &mut self,
+        path: &Path,
+    ) -> Result<BTreeMap<PackageName, UnresolvedPackageGroup>> {
```

Comment:
Could this preserve the `Vec` return value? This looks like it's returning an error on duplicate packages but most of the rest of WIT processing ends up merging things together so that may be best to do here instead.